modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -145,13 +156,266 @@ class SoftMaxSubgraph : public Subgraph
+
+        setFusedNode("Scales", input);
+    }
+
+    virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
+                    std::vector<int>& matchedNodesIds,
+                    std::vector<int>& targetNodesIds) CV_OVERRIDE
+    {
+        if (FuseUpsampleSubgraph::match(net, nodeId, matchedNodesIds, targetNodesIds))
+        {
+            std::vector<cv::Mat> nodes_attributes;
+            for (int i = matchedNodesIds.front(); i < matchedNodesIds.back(); i++)
+            {
+                if (std::find(matchedNodesIds.begin(), matchedNodesIds.end(), i) == matchedNodesIds.end())
+                {
+                    opencv_onnx::NodeProto* constant_node = net->getNode(i).dynamicCast<ONNXNodeWrapper>()->node;
+                    opencv_onnx::TensorProto tensor_proto = constant_node->attribute(0).t();
+                    nodes_attributes.push_back(getMatFromTensor(tensor_proto));
+                }
+            }
+            CV_Assert(nodes_attributes[0].total() == 1/*indexH*/ && nodes_attributes[1].total() == 1/*scaleH*/ &&
+                nodes_attributes[2].total() == 1/*indexW*/ && nodes_attributes[3].total() == 1/*scaleW*/);
+
+            if (nodes_attributes[0].at<int>(0) == 2 && nodes_attributes[2].at<int>(0) == 3)
+            {
+                scaleH = (int)nodes_attributes[1].at<float>(0);
+                scaleW = (int)nodes_attributes[3].at<float>(0);
+            }
+            else
+            {
+                scaleH = (int)nodes_attributes[3].at<float>(0);
+                scaleW = (int)nodes_attributes[1].at<float>(0);
+            }
+            return true;
+        }
+        return false;
+    }
+};
+
+class UpsampleSubgraph : public FuseUpsampleSubgraph
+{
+public:
+    UpsampleSubgraph()
+    {
+        int input = addNodeToMatch("");
+
+        int scaleHW = addNodeToMatch("Scales", input);
+        int scaleNC = addNodeToMatch("Constant");
+        int cast1 = addNodeToMatch("Cast", scaleHW);
+
+        int shape = addNodeToMatch("Shape", input);
+        int slice = addNodeToMatch("Slice", shape);
+        int cast2 = addNodeToMatch("Cast", slice);
+
+        int inpDiv = addNodeToMatch("Div", cast1, cast2);
+        int inpConcat2 = addNodeToMatch("Concat", scaleNC, inpDiv);
+
+        addNodeToMatch("Upsample", input, inpConcat2);
+        setFusedNode("Upsample", input);
+    }
+
+    virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
+                       std::vector<int>& matchedNodesIds,
+                       std::vector<int>& targetNodesIds) CV_OVERRIDE
+    {
+        if (FuseUpsampleSubgraph::match(net, nodeId, matchedNodesIds, targetNodesIds))
+        {
+            Ptr<ImportNodeWrapper> scales = net->getNode(matchedNodesIds[0]);
+            opencv_onnx::NodeProto* node = scales.dynamicCast<ONNXNodeWrapper>()->node;
+            opencv_onnx::AttributeProto attr = node->attribute(1);
+            CV_Assert(attr.name() == "scales");
```

Comment:
Let's not force this name - we know that scales are located in this node and the name should not block as to extract them.

Reply:
Okay, I will change that.